paper/VIT/train_memory_vit.py

```diff
@@ -0,0 +1,245 @@
+        if hparams.cutmix:
+            self.cutmix = CutMix(hparams.size, beta=1.)
+        if hparams.mixup:
+            self.mixup = MixUp(alpha=1.)
+        self.log_image_flag = hparams.api_key is None
+
+    def forward(self, x,ss):
+        return self.model(x,ss)
+
+    def configure_optimizers(self):
+        self.optimizer = torch.optim.Adam(self.model.parameters(), lr=self.hparams.lr, betas=(self.hparams.beta1, self.hparams.beta2), weight_decay=self.hparams.weight_decay)
+        self.base_scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(self.optimizer, T_max=self.hparams.max_epochs, eta_min=self.hparams.min_lr)
+        self.scheduler = warmup_scheduler.GradualWarmupScheduler(self.optimizer, multiplier=1., total_epoch=self.hparams.warmup_epoch, after_scheduler=self.base_scheduler)
+        return [self.optimizer], [self.scheduler]
+
+    def training_step(self, batch, batch_idx):
+
+        img, label = batch['data']
+        memory_input, memory_label = batch['memory']
+        if self.hparams.cutmix or self.hparams.mixup:
+            if self.hparams.cutmix:
+                img, label, rand_label, lambda_= self.cutmix((img, label))
+            elif self.hparams.mixup:
+                if np.random.rand() <= 0.8:
+                    img, label, rand_label, lambda_ = self.mixup((img, label))
+                else:
+                    img, label, rand_label, lambda_ = img, label, torch.zeros_like(label), 1.
+            out = self.model(img,memory_input)
+            loss = self.criterion(out, label)*lambda_ + self.criterion(out, rand_label)*(1.-lambda_)
+        else:
+            out = self(img,memory_input)
+            loss = self.criterion(out, label)
+
+        if not self.log_image_flag and not self.hparams.dry_run:
+            self.log_image_flag = True
+            self._log_image(img.clone().detach().cpu())
+
+        acc = torch.eq(out.argmax(-1), label).float().mean()
+        self.log("loss", loss)
+        self.log("acc", acc)
+        return loss
+
+    def training_epoch_end(self, outputs):
+        self.log("lr", self.optimizer.param_groups[0]["lr"])
+
+    def validation_step(self, batch, batch_idx):
+        
+        img, label = batch['data']
+        memory_input, memory_label = batch['memory']
+
+        out = self(img,memory_input)
+        loss = self.criterion(out, label)
+        acc = torch.eq(out.argmax(-1), label).float().mean()
+        self.log("val_loss", loss)
+        self.log("val_acc", acc)
+        return loss
+
+
+    # def train_dataloader(self):
+    #     loaders = {"data": train_dl, "memory": mem_loader}
+    #     combined_loaders = CombinedLoader(loaders, "max_size_cycle")
+    #     return combined_loaders
+    
+    # def val_dataloader(self):
+    #     loaders = {"data": test_dl, "memory": mem_loader}
+    #     combined_loaders = CombinedLoader(loaders, "max_size_cycle")
+    #     return combined_loaders
+
+
+    def _log_image(self, image):
```

Comment:
Looks like this is the only file where we do not use typing. Is there a reason?

P.S. not for now, but a todo: set up type checking with MyPy

Reply:
Because it is the https://github.com/omihub777/ViT-CIFAR/blob/main/main.py file, slightly changed only for working with Memory Wrap. Added the reference in [0ad70d5](https://github.com/KRLGroup/memory-wrap/pull/3/commits/0ad70d5920298e86ff048a2e3d9513b3a4310708)